### PR TITLE
Link Integrity

### PR DESCRIPTION
**Pull request by danielamormocea**
Solves #2396

Requires plone.restapi (https://github.com/plone/plone.restapi/pull/953)

image

### CODE REVIEW

File: src/components/manage/Contents/Contents.jsx
Comment (avoinea) [inline src/components/manage/Contents/Contents.jsx:1166]: @danielamormocea Please use i18n messages here.

File: src/actions/content/content.js
Comment (avoinea) [inline src/actions/content/content.js:16]: @danielamormocea Maybe we can find a better name, more self-explenatory, for this action.

**Review by avoinea — CHANGES_REQUESTED** [review]

**Comment by sneridagh** [comment]
It seems we already had a rotten PR going on: https://github.com/plone/volto/pull/2783 😫 that did the same thing.

@danielamormocea could you take a look and try to merge them (if there's something valuable, that I'm not aware now, i18n maybe?) there that we could reuse in this one? I'm biassed to use this one though, make sure that the i18n is also in sync.

As soon as we have the PR merged in p.restapi we can merge this one.

**Review by avoinea — APPROVED** [review]

**Comment by cekk** [comment]
@danielamormocea @avoinea we changed the endpoint method from POST to GET, so i think that you have to change something in the frontend call

**Comment by sneridagh** [comment]
@danielamormocea I think they are changing the backend PR a bit. Please talk to @cekk to adjust

**Review by sneridagh — APPROVED** [review]
LGTM, when the restapi is merged we can merge this.

**Comment by tisto** [comment]
@danielamormocea awesome work! I just added the plone.restapi checkout to this PR. I would love to see a Cypress test that covers the basic use cases:

- delete object that is referenced by a text block link
- delete object that is referenced by an attribute reference
- delete parent object of an object that is referenced by a link (either text or attribute or maybe even both)

**Comment by tisto** [comment]
@danielamormocea the link integrity feature works very well. I'd just have a few suggestions for the UX:

Contents (1)

**Comment by tisto** [comment]
@danielamormocea nested folder structures are taken into account, which is good. Though, the system currently does not show the right breaches:

Contents (2)

**Comment by danielamormocea** [comment]
> @danielamormocea the link integrity feature works very well. I'd just have a few suggestions for the UX:
> 
> Contents (1)

In terms of UI, I have no better idea than to simply mention again before listing the items, that those are going to be deleted (although it seems redundant), or to add a separator again after the first list of items (but for me that looks uglier).

Furthermore, regarding the breaches, if there is a page inside a page that is referenced somewhere, the response won't give me that information, so the main page (the one selected) is the only "reference" I get, therefore the only page I can show to the user.

**Comment by danielamormocea** [comment]
@tisto I also added the test for the reference, but unfortunately the linkintegrity api part does not detect objects referenced by a text block link

**Comment by tisto** [comment]
@danielamormocea I was suspecting that Plone and the restapi are the issue here. Then I guess we have to specify in plone.restapi how the backend is supposed to work and then work on the implementation there.

**Comment by tisto** [comment]
@danielamormocea would you be able to create a PR in plone.restapi documenting the missing functionality in a Python unit test?

FYI: Acceptance tests and code analysis are failing for this PR.

**Comment by danielamormocea** [comment]
@tisto  Acceptance tests are failing because there is no linkintegrity api for deleting items from contents (the failing tests are in content.js and folder-contents.js)

**Review by rohnsha0 — APPROVED** [review]

**Comment by sneridagh** [comment]
@danielamormocea I hope that now that we have the backend side, we are green and ready to merge.

**Comment by sneridagh** [comment]
@danielamormocea we have a test that fails, could you please take a look? Thanks!

**Comment by sneridagh** [comment]
@danielamormocea I fixed the Cypress test (it was using draftJS still). 
However, I can't see the @linkintegrity call happening... Could you please take a look?

I'd like to have this for Volto 16 which will happen next week for sure.

/cc @avoinea @alecghica

**Comment by sneridagh** [comment]
@avoinea I still couldn't take a look at this, it should work now but it keeps failing. Should be a silly thing.

**Pull request merged**
